### ISSUE TITLE
java.util.ConcurrentModificationException Tinker's Construct

### ISSUE THREAD

**Issue by peterodavies27atgmailcom** [MultiMC/MultiMC5#3439]
System Information
-----------------------------
MultiMC version: MultiMC 5 - Version 0.6.11-1430 on win32

Operating System: Windows 10

Summary of the issue or suggestion: MC crashes on launch
----------------------------------------------

What should happen: MC shouldn't crash obviously
------------------------------

Steps to reproduce the issue (Add more if needed):
-------------------------------------------------------------
1. Launch MC

2. Wait for it to load (very large modpack)

Suspected cause: VanillaFix says it's TiCon but it doesn't tell anything about which mod right above the crash report
---------------------------

Logs/Screenshots: https://paste.ee/p/PRPHC
----------------------------
[//]: # (Please refer to https://github.com/MultiMC/MultiMC5/wiki/Log-Upload for instructions on how to attach your logs.)

Additional Info: ???
---------------------------

**Comment by peterix** [comment]
```
java.util.ConcurrentModificationException
    at java.util.HashMap$HashIterator.nextNode(HashMap.java:1445)
    at java.util.HashMap$KeyIterator.next(HashMap.java:1469)
    at slimeknights.tconstruct.library.client.CustomTextureCreator.createMaterialTextures(CustomTextureCreator.java:110)****
```

Contact whoever maintains that and make them fix their bugs.

**Comment by QuImUfu** [comment]
@peterodavies27atgmailcom this is a Tinkers Construct bug that AFAIK will never be fixed as Tinkers Construct in 1.12.2 is unmaintained.  
A diff badly working arround the issue without fixing the root cause ( i am prety sure i looked at the cause when i wrote this, and this should not have bad side effects, a proper fix would have been rather complicated and would have changed nothing about the result) :  
```diff
diff --git a/build.properties b/build.properties
index 6bb6644c1..716026d7f 100644
--- a/build.properties
+++ b/build.properties
@@ -1,4 +1,4 @@
-mod_version=2.13.0
+mod_version=2.13.1
 minecraft_version=1.12.2
 minecraft_version_short=1.12
 forge_version=14.23.5.2831
diff --git a/src/main/java/slimeknights/tconstruct/library/client/CustomTextureCreator.java b/src/main/java/slimeknights/tconstruct/library/client/CustomTextureCreator.java
index 9056a4875..9a3cc4600 100644
--- a/src/main/java/slimeknights/tconstruct/library/client/CustomTextureCreator.java
+++ b/src/main/java/slimeknights/tconstruct/library/client/CustomTextureCreator.java
@@ -27,6 +27,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import java.util.Arrays;
 
 import javax.annotation.Nonnull;
 
@@ -107,7 +108,15 @@ public class CustomTextureCreator implements IResourceManagerReloadListener {
 
   private void createMaterialTextures(TextureMap map) {
     // Create textures for toolparts and tools - Textures that need 1 per material
-    for(ResourceLocation baseTexture : baseTextures) {
+    Set<ResourceLocation> done = Sets.newHashSet();
+    int tries = 0;
+    do {
+      Set<ResourceLocation> baseTexturesCopy = Sets.newHashSet(baseTextures);
+      for(ResourceLocation baseTexture : baseTexturesCopy) {
+        if(done.contains(baseTexture)){
+          continue;
+        }
+        done.add(baseTexture);
        // exclude missingno :I
        if(baseTexture.toString().equals("minecraft:missingno")) {
                continue;
@@ -145,6 +154,8 @@ public class CustomTextureCreator implements IResourceManagerReloadListener {
        
        sprites.put(baseTexture.toString(), builtSprites);
       }
+      log.info(++tries + ". Loop done!");
+       } while (!done.containsAll(baseTextures));
   }
 
   private TextureAtlasSprite createTexture(Material material, ResourceLocation baseTexture, TextureMap map) {
```  
[here](https://dl.dropbox.com/s/2f3p9x79k8euq0l/TConstruct-1.12.2-2.13.1.DEV.1483c3c08.jar) is a Tcon build with that patch applied.

**Comment by peterodavies27atgmailcom** [comment]
oh, thanks

**Comment by DabbingEevee** [comment]
> @peterodavies27atgmailcom this is a Tinkers Construct bug that AFAIK will never be fixed as Tinkers Construct in 1.12.2 is unmaintained. A diff badly working arround the issue without fixing the root cause ( i am prety sure i looked at the cause when i wrote this, and this should not have bad side effects, a proper fix would have been rather complicated and would have changed nothing about the result) :
> 
> ```diff
> diff --git a/build.properties b/build.properties
> index 6bb6644c1..716026d7f 100644
> --- a/build.properties
> +++ b/build.properties
> @@ -1,4 +1,4 @@
> -mod_version=2.13.0
> +mod_version=2.13.1
>  minecraft_version=1.12.2
>  minecraft_version_short=1.12
>  forge_version=14.23.5.2831
> diff --git a/src/main/java/slimeknights/tconstruct/library/client/CustomTextureCreator.java b/src/main/java/slimeknights/tconstruct/library/client/CustomTextureCreator.java
> index 9056a4875..9a3cc4600 100644
> --- a/src/main/java/slimeknights/tconstruct/library/client/CustomTextureCreator.java
> +++ b/src/main/java/slimeknights/tconstruct/library/client/CustomTextureCreator.java
> @@ -27,6 +27,7 @@ import java.util.List;
>  import java.util.Map;
>  import java.util.Optional;
>  import java.util.Set;
> +import java.util.Arrays;
>  
>  import javax.annotation.Nonnull;
>  
> @@ -107,7 +108,15 @@ public class CustomTextureCreator implements IResourceManagerReloadListener {
>  
>    private void createMaterialTextures(TextureMap map) {
>      // Create textures for toolparts and tools - Textures that need 1 per material
> -    for(ResourceLocation baseTexture : baseTextures) {
> +    Set<ResourceLocation> done = Sets.newHashSet();
> +    int tries = 0;
> +    do {
> +      Set<ResourceLocation> baseTexturesCopy = Sets.newHashSet(baseTextures);
> +      for(ResourceLocation baseTexture : baseTexturesCopy) {
> +        if(done.contains(baseTexture)){
> +          continue;
> +        }
> +        done.add(baseTexture);
>         // exclude missingno :I
>         if(baseTexture.toString().equals("minecraft:missingno")) {
>                 continue;
> @@ -145,6 +154,8 @@ public class CustomTextureCreator implements IResourceManagerReloadListener {
>         
>         sprites.put(baseTexture.toString(), builtSprites);
>        }
> +      log.info(++tries + ". Loop done!");
> +       } while (!done.containsAll(baseTextures));
>    }
>  
>    private TextureAtlasSprite createTexture(Material material, ResourceLocation baseTexture, TextureMap map) {
> ```
> 
> [here](https://dl.dropbox.com/s/2f3p9x79k8euq0l/TConstruct-1.12.2-2.13.1.DEV.1483c3c08.jar) is a Tcon build with that patch applied.

thank you so much
